### PR TITLE
fix: value() might also return undefined if no value is expected

### PR DESCRIPTION
**Pull request by danielpeintner**
fixes https://github.com/eclipse/thingweb.node-wot/issues/747

### CODE REVIEW

**Comment by egekorkan** [comment]
> Even if I agree that these changes are a solution for https://github.com/eclipse/thingweb.node-wot/issues/747, I am not convinced that assuming an undefined schema is equal to "no-content". Maybe I missed some discussion on the TD repository, do you reach a resolution on this problem? 
> 
> 
> 
> Why don't we try to parse the value and return what the protocol is giving us? does it create problems down the road? 

I agree with this approach as well. There can be different valid cases of not defining a schema:
- laziness of the Thing developer (which should penalise the Consumer developer any further by node-wot)
- the schema defined by `additionalResponses` that node-wot does not support yet (I think)
- The payload is not describable by a JSON Schema due to different payload formats
- Similar to the above, the payload is described by terms in the forms

**Comment by danielpeintner** [comment]
Valid points. I updated the PR so that
* IF there is a schema it will be used to validate
* IF there is *no* schema it tries to report what the binding gives us (or `undefined`)

File: packages/core/src/interaction-output.ts [hunk 1]
```diff
@@ -63,35 +63,34 @@ export class InteractionOutput implements WoT.InteractionOutput {
...
         }
-        // read fully the stream
 
+        // read fully the stream
         const data = await ProtocolHelpers.readStreamFully(this.content.body);
         this.dataUsed = true;
         this.buffer = data;
+
         // call the contentToValue
         const value = ContentSerdes.get().contentToValue({ type: this.content.type, body: data }, this.schema);
```
Comment (relu91) [inline packages/core/src/interaction-output.ts:78]: One concern: if `this.schema` is nullable but in `contentToValue` it is not... how TS compiler accept this call without warning you?

**Review by relu91 — APPROVED** [review]
As I expressed below there might be some problems. For example, some binding uses the schema to parse the payload correctly... Anyhow, I think we can create an issue and fix it later!

File: packages/core/src/interaction-output.ts (hunk 1 above)
Comment (danielpeintner) [reply packages/core/src/interaction-output.ts:78]: No TS compiler warning but I don't think the compiler can figure it out in this case..

We might want to add a note to the `ContentSerdes` interface methods a la 
"Note: schema parameter might be null/undefined"

https://github.com/eclipse/thingweb.node-wot/blob/3803354edafd9def2f90553f7ce9ef2eeba37ab0/packages/core/src/content-serdes.ts#L29-L30

The alternative would be to change the signature to mark schema as optional.

`bytesToValue(bytes: Buffer, schema?: DataSchema, parameters? ...)`

Not sure though whether changing the interface (at this point in time) makes sense. Any proposal? 

Question: What is the *best* way to go
1. Just add a note without changing the signature
2. change signature of interface
3. Do 2. it in a follow-up PR (with other breaking changes)

Note: I can check the current implementations (and update the PR accordingly) whether they properly deal with schema being null/undefined. There are some that fail like OctetstreamCodec (need to check others).

Comment (JKRhb) [reply packages/core/src/interaction-output.ts:78]: Making the schema optional sounds like a good idea, I also implemented it this way in `dart_wot` (see [here](https://github.com/namib-project/dart_wot/blob/812bddc568be58d4566acbbeb6f6b2cf66a93ad5/lib/src/core/content_serdes.dart#L211)). If no schema is passed, then the payload is simply returned and not validated.

Comment (danielpeintner) [reply packages/core/src/interaction-output.ts:78]: Makes sense. I created a new issue https://github.com/eclipse/thingweb.node-wot/issues/753 and I suggest we deal with it in a follow-up PR...

**Pull request merged**